File: app/src/main/java/com/jorbital/gymstat/utils/LiveRealmData.java
package com.jorbital.gymstat.utils;

import android.arch.lifecycle.LiveData;

import io.realm.RealmChangeListener;
import io.realm.RealmObject;
import io.realm.RealmResults;

/**
 * This is a wrapper for the RealmResults to expose them as Lifecycle aware LiveData.
 */

public class LiveRealmData<T extends RealmObject> extends LiveData<RealmResults<T>> {

    private RealmResults<T> results;
    private final RealmChangeListener<RealmResults<T>> listener =
            new RealmChangeListener<RealmResults<T>>() {
                @Override
                public void onChange(RealmResults<T> results) { setValue(results);}
            };

    public LiveRealmData(RealmResults<T> realmResults) {
        results = realmResults;
    }

    @Override
    protected void onActive() {
        results.addChangeListener(listener);
    }

    @Override
    protected void onInactive() {
        results.removeChangeListener(listener);
    }
}

File: app/src/main/java/com/jorbital/gymstat/utils/LiveRealmData.kt
package com.jorbital.gymstat.utils

import android.arch.lifecycle.LiveData

import io.realm.RealmChangeListener
import io.realm.RealmObject
import io.realm.RealmResults

/**
 * This is a wrapper for the RealmResults to expose them as Lifecycle aware LiveData.
 */

class LiveRealmData<T : RealmObject>(var results: RealmResults<T>) : LiveData<RealmResults<T>>() {
    private val listener = RealmChangeListener<RealmResults<T>> { results -> value = results }

    override fun onActive() {
        results.addChangeListener(listener)
    }

    override fun onInactive() {
        results.removeChangeListener(listener)
    }
}
